bin/worker.ts

```diff
@@ -1,65 +1,105 @@
 import { ReadStream, createReadStream } from 'fs';
+import { stat } from 'node:fs/promises';
+import { basename, join } from 'node:path';
 import retryPromise from 'promise-retry';
-import { SymbolFileInfo, SymbolFileType } from './symbol-file-info';
+import { WorkerPool, cpus } from 'workerpool';
+import { SymbolFileInfo } from './info';
+import { tmpDir } from './tmp';
+const workerCount = cpus;
 
-export function createWorkersFromSymbolFiles(symbolFiles: SymbolFileInfo[], workerCount: number, clients: [SymbolsApiClient, VersionsApiClient]): Array<UploadWorker> {
+export function createWorkersFromSymbolFiles(workerPool: WorkerPool, symbolFiles: SymbolFileInfo[], clients: [SymbolsApiClient, VersionsApiClient]): Array<UploadWorker> {
     const numberOfSymbols = symbolFiles.length;
     
     if (workerCount >= numberOfSymbols) {
-        return symbolFiles.map((symbolFile, i) => new UploadWorker(i + 1, [symbolFile], ...clients));
+        return symbolFiles.map((symbolFile, i) => new UploadWorker(i + 1, [symbolFile], workerPool, ...clients));
     }
 
-    const sorted = symbolFiles.sort((a, b) => a.size - b.size);
-    const symbolFilesChunks = splitToChunks(sorted, workerCount);
-    return symbolFilesChunks.map((chunk, i) => new UploadWorker(i + 1, chunk, ...clients));
-}
-
-function splitToChunks<T>(array: Array<T>, parts: number): Array<Array<T>> {
-    const copy = [...array];
-    const result = [] as Array<Array<T>>;
-    for (let i = parts; i > 0; i--) {
-        result.push(copy.splice(0, Math.ceil(array.length / parts)));
-    }
-    return result;
+    const symbolFilesChunks = splitToChunks(symbolFiles, workerCount);
+    return symbolFilesChunks.map((chunk, i) => new UploadWorker(i + 1, chunk, workerPool, ...clients));
 }
 
 export class UploadWorker {
     private createReadStream = createReadStream;
     private retryPromise = retryPromise;
+    private stat = stat;
     private toWeb = ReadStream.toWeb;
 
     constructor(
         public readonly id: number,
         public readonly symbolFileInfos: SymbolFileInfo[],
-        private symbolsClient: SymbolsApiClient,
-        private versionsClient: VersionsApiClient,
+        private readonly pool: WorkerPool,
+        private readonly symbolsClient: SymbolsApiClient,
+        private readonly versionsClient: VersionsApiClient,
     ) { }
     
     async upload(database: string, application: string, version: string): Promise<void> {
         console.log(`Worker ${this.id} uploading ${this.symbolFileInfos.length} symbol files...`);
 
         for (const symbolFileInfo of this.symbolFileInfos) {
             await this.retryPromise(
-                (retry) => this.uploadSingle(database, application, version, symbolFileInfo).catch(retry)
+                (retry) => this.uploadSingle(database, application, version, symbolFileInfo)
+                    .catch((error) => {
+                        retry(error) // TODO BG remove
```

Comment:
Oops!

Reply:
With all these changes, I think this retry should be moved to just wrap the upload portion...

Reply:
Fixed in https://github.com/BugSplat-Git/symbol-upload/pull/72/commits/37e68ab201d1b48ec148045b71d2d83a45e4e199